### DISCUSSION TITLE
Sort posts by title in a tag page

### DISCUSSION THREAD

**Discussion by Julianoe** [11ty/eleventy#2557] [Q&A] [answered]
Hey, it's me again, trying to figure all this thing out :)
I followed some tips and advice to build a tag "archive" page like in the [quick tip #004](https://www.11ty.dev/docs/quicktips/tag-pages/).
Now what i'm trying to do is make this page display posts in a A-Z order (it's by date by default).

I tried to imitate several people ([here for ex](https://github.com/11ty/eleventy/issues/898#issuecomment-662290629)) that built a filter that uses localeCompare but it throws an error I can't manage to understand: `TypeError: Cannot read properties of undefined (reading 'localeCompare') (via Template render error)`.
```js
  eleventyConfig.addFilter('sortByTitle', values => {
      return values.slice().sort((a, b) => a.data.title.localeCompare(b.data.title))
  })
```

Then I tried to do it another way, like @pdehaan suggested [here](https://github.com/11ty/eleventy/issues/898#issuecomment-897071947), by using the default Sort() function fron Nunjuck but to no success as it returns an other error `TypeError: Cannot convert undefined or null to object (via Template render error)`. I only changed that line:
```njk
{% for note in taglist | sort(false, false, 'data.title') %}
```

I guess there is something I don't understand about the way data is passed around but can't find anymore info to do what I'm trying to achieve. 

Here is the code from tags.njk:
```njk
---
layout: default
pagination:
  data: collections
  size: 1
  alias: tag
permalink: /tags/{{ tag }}/
---

<h1>Étiquette: « {{ tag }} »</h1>

<ul>
{% set taglist = collections[ tag ] %}
{% for note in taglist | sortByTitle %}
  <li><a href="{{ note.url }}">{{ note.data.title }}</a> {%if note.data.date %}<span class="small">| <time>{{ note.data.date | dateToLocalString }}</time></span>{% endif %}</li>
{% endfor %}
</ul>
```

**Comment by pdehaan** [comment, marked as answer]
Your logic here looks right to me:
```js
  eleventyConfig.addFilter('sortByTitle', values => {
      return values.slice().sort((a, b) => a.data.title.localeCompare(b.data.title))
  })
```

&hellip; but this error message seems to imply that one template might not have a `title` attribute (it seems to be `undefined` causing an error trying to then read `undefined.localeCompare()`):

> _"TypeError: Cannot read properties of `undefined` (reading 'localeCompare') (via Template render error)"_

**Reply by pdehaan** [reply]
I was able to reproduce the error locally if I renamed one of my sample page's frontmatter from `title` to `title2`, which causes `title` to be `undefined` and throw the error.
Possible workarounds are (in no particular order):
- casting to String: `.sort((a, b) => String(a.data.title).localeCompare(b.data.title))` 
- defaulting to an empty string if `a.data.title` is falsy: `.sort((a, b) => (a.data.title || "").localeCompare(b.data.title))`
- [chaining operator (`?.`)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) to guard against falsy values: `.sort((a, b) => a.data.title?.localeCompare(b.data.title))`

But my guess is that it's bad input and one of the pages is missing a `title` value.

**Reply by Julianoe** [reply]
> But my guess is that it's bad input and one of the pages is missing a title value.

I admit I did not think of this simple possibility. I will check my dataset and try to fix all this with your advice.
Thanks a lot for all the answers, you help is very much appreciated 🙏

**Reply by pdehaan** [reply]
In the past I've used tools like <kbd>grep</kbd> or <kbd>git grep</kbd> to try and find files without matching strings to try and find these types of errors (but your mileage may vary):

```sh
grep -L "title:" ./src/*.njk
git grep -L "title:" src
```

(where the `-L` flag lists files without matches; see https://til.hashrocket.com/posts/5229cacf56-grep-for-files-without-a-match)

**Reply by Julianoe** [reply]
Yes I had my own little bash script to detect this kind of things. I've cleaned up my dataset. Every file now includes a title, correctly formatted.

I've tried every one of the 3 solutions you suggested and they all return this same error: 

```
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] 1. Having trouble rendering njk template ./tags.njk (via TemplateContentRenderError)
[11ty] 2. (./tags.njk)
[11ty]   TypeError: Cannot read properties of undefined (reading 'title') (via Template render error)
[11ty] 
[11ty] Original error stack trace: Template render error: (./tags.njk)
[11ty]   TypeError: Cannot read properties of undefined (reading 'title')
[11ty]     at Object._prettifyError (/home/user/eleventy-garden/node_modules/nunjucks/src/lib.js:36:11)
```

**Reply by pdehaan** [reply]
Not sure.

Somewhere in tags.njk (or maybe some filter/shortcode called from that page) it looks like you're trying to access `.title` from an undefined property.

> _"[11ty]   TypeError: Cannot read properties of `undefined` (reading `'title'`) (via Template render error)"_

What does your tags.njk look like?

**Reply by Julianoe** [reply]
I have currently 2 templates listing notes:

`all-notes.njk` : on this one the order works, but a (weird?) thing seems to be involved: they are sorted by title like my `sort(false, false, 'data.title')` asks it to do, but posts that have no date: specified are listed after. I have then  2 lists: first list of posts _that have dates_ sorted by title A-Z, then list of posts _with no date_ by title A-Z. I don't know how to avoid the date being factored into the sorting.

```js
---
layout: default
title: "All notes"
description: "List of all notes sorted by title."
---
<ul>
{% for p in collections.notes | sort(false, false, 'data.title') %}
  <li><a href="{{ p.url | url }}" >{{ p.data.title }}</a> <span class="small">{{ p.data.date | dateToLocalString }}</span></li>
{% endfor %}
</ul>
```
`tags.njk` : where I tried the things you advised me to do and could not manage to make it work. I either get the error mentioned previously for any of the 3 solutions you proposed to put into the sortByTitle filter, or either get the error mentioned below (

```js
---
layout: default
title: "Tag"
description: "Page listing all notes for one tag"
pagination:
  data: collections
  size: 1
  alias: tag
permalink: /tags/{{ tag }}/
---
<h1>Étiquette: « {{ tag }} »</h1>
<ul>
{% set taglist = collections[ tag ] %}
{% for p in taglist | sort(false, false, 'p.data.title') %}
{# {% for p in taglist | sortByTitle %} #}
  <li><a href="{{ p.url }}">{{ p.data.title }}</a> {%if p.data.date %}<span class="small">| <time>{{ p.data.date | dateToLocalString }}</time></span>{% endif %}</li>
{% endfor %}
</ul>
```
`{% for p in taglist | sort(false, false, 'p.data.title') %}` (or 'data.title' only) gives me this error, when it worked in all-notes.njk:

```bash
1. Having trouble rendering njk template ./tags.njk (via TemplateContentRenderError)
[11ty] 2. (./tags.njk)
[11ty]   TypeError: Cannot convert undefined or null to object (via Template render error)
[11ty] 
[11ty] Original error stack trace: Template render error: (./tags.njk)
[11ty]   TypeError: Cannot convert undefined or null to object
```

for reference I add below the `sortByTitle` filter from `.eleventy.js`: 

```js
    /* throws errors : https://github.com/11ty/eleventy/discussions/2557#discussioncomment-3620503 */
    eleventyConfig.addFilter('sortByTitle', values => {

        // return values.slice().sort((a, b) => a.data.title.localeCompare(b.data.title))
        // return values.slice().sort((a, b) => String(a.data.title).localeCompare(b.data.title)) // casting to String
        // return values.slice().sort((a, b) => (a.data.title || "").localeCompare(b.data.title)) // defaulting to an empty string if a.data.title is falsy:
        return values.slice().sort((a, b) => a.data.title?.localeCompare(b.data.title)) // chaining operator (?.) to guard against falsy values: 
    })
```

**Reply by pdehaan** [reply]
Not sure I'm seeing the same behavior for your first point:
```njk
---
# src/notes.njk
# layout: default
title: "All notes"
description: "List of all notes sorted by title."
---

<ul>
{% for p in collections.notes | sort(false, false, 'data.title') %}
  <li><a href="{{ p.url | url }}" >{{ p.data.title }}</a> <span class="small">{{ p.data.date }}</span></li>
{% endfor %}
</ul>
```

### OUTPUT
```html
<ul>
  <li><a href="/notes/five/" >FiVe</a> <span class="small">Wed May 14 2003 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</span></li>
  <li><a href="/notes/four/" >FoUr</a> <span class="small">Sun Apr 13 2014 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</span></li>
  <li><a href="/notes/one/" >oNe</a> <span class="small">Fri Jan 30 2009 16:00:00 GMT-0800 (Pacific Standard Time)</span></li>
  <li><a href="/notes/three/" >ThReE</a> <span class="small"></span></li>
  <li><a href="/notes/two/" >tWo</a> <span class="small">Wed Sep 18 2019 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</span></li>
</ul>
```

```sh
> eleventy

[11ty] Writing www/notes/five/index.html from ./src/notes/five.njk
[11ty] Writing www/notes/one/index.html from ./src/notes/one.njk
[11ty] Writing www/notes/four/index.html from ./src/notes/four.njk
[11ty] Writing www/notes/two/index.html from ./src/notes/two.njk
[11ty] Writing www/notes/index.html from ./src/notes.njk
[11ty] Writing www/notes/three/index.html from ./src/notes/three.njk
[11ty] Wrote 6 files in 0.04 seconds (v1.0.2)
```

My notes are sorted by `title`, not by `date` (or grouped by absence of a date)

**Reply by pdehaan** [reply]
And do your second point, your code seems to work locally for me:

```njk
---
# src/tags.njk
# layout: default
title: "Tag"
description: "Page listing all notes for one tag"
pagination:
  data: collections
  size: 1
  alias: tag
permalink: /tags/{{ tag | slugify }}/
---
<h1>Étiquette: « {{ tag }} »</h1>
<ul>
{% set taglist = collections[ tag ] %}
{# {% for p in taglist | sort(false, false, 'p.data.title') %} #}
{% for p in taglist | sortByTitle %}
  <li data-tags="{{ p.data.tags }}"><a href="{{ p.url }}">{{ p.data.title }}</a> {%if p.data.date %}<span class="small">| <time>{{ p.data.date }}</time></span>{% endif %}</li>
{% endfor %}
</ul>
```

### OUTPUT

```html
<!-- www/tags/all/index.html --->
<h1>Étiquette: « all »</h1>
<ul>
  <li data-tags=""><a href="/notes/">All notes</a> </li>
  <li data-tags="notes,animal"><a href="/notes/five/">FiVe</a> <span class="small">| <time>Wed May 14 2003 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,vegetable"><a href="/notes/four/">FoUr</a> <span class="small">| <time>Sun Apr 13 2014 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,mineral"><a href="/notes/one/">oNe</a> <span class="small">| <time>Fri Jan 30 2009 16:00:00 GMT-0800 (Pacific Standard Time)</time></span></li>
  <li data-tags=""><a href="/tags/all/">Tag</a> </li>
  <li data-tags="notes,animal,mineral"><a href="/notes/three/">ThReE</a> </li>
  <li data-tags="notes,vegetable,til"><a href="/notes/two/">tWo</a> <span class="small">| <time>Wed Sep 18 2019 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,til"><a href="/notes/six/"></a> </li>
  <li data-tags="notes"><a href="/notes/seven/"></a> </li>
  <li data-tags="notes"><a href="/notes/eight/"></a> </li>
</ul>
```

Using your custom `sortByTitle` filter:
```js
  eleventyConfig.addFilter('sortByTitle', (values) => {
    // chaining operator (?.) to guard against falsy values: 
    return [...values].sort((a, b) => a.data.title?.localeCompare(b.data.title));
  });
```

**Reply by pdehaan** [reply]
Note that trying to call [`sort`](https://mozilla.github.io/nunjucks/templating.html#sort-arr-reverse-casesens-attr) on `"p.data.title"` doesn't work, and you need to use `"data.title"` like you did in all-notes.njk.

```diff
-{% for p in taglist | sort(false, false, 'p.data.title') %}
+{% for p in taglist | sort(false, false, 'data.title') %}
```

Using `p.data.title` didn't throw any errors, it just didn't sort the results (presumably because every result would sort by `undefined` or something, returning the original unsorted order).

### Using `{% for p in taglist | sort(false, false, 'p.data.title') %}`
```html
<h1>Étiquette: « all »</h1>
<ul>
  <li data-tags="notes,animal"><a href="/notes/five/">FiVe</a> <span class="small">| <time>Wed May 14 2003 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,mineral"><a href="/notes/one/">oNe</a> <span class="small">| <time>Fri Jan 30 2009 16:00:00 GMT-0800 (Pacific Standard Time)</time></span></li>
  <li data-tags="notes,vegetable"><a href="/notes/four/">FoUr</a> <span class="small">| <time>Sun Apr 13 2014 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,vegetable,til"><a href="/notes/two/">tWo</a> <span class="small">| <time>Wed Sep 18 2019 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,animal,mineral"><a href="/notes/three/">ThReE</a> </li>
  <li data-tags=""><a href="/notes/">All notes</a> </li>
  <li data-tags=""><a href="/tags/all/">Tag</a> </li>
  <li data-tags="notes,til"><a href="/notes/six/"></a> </li>
  <li data-tags="notes"><a href="/notes/seven/"></a> </li>
  <li data-tags="notes"><a href="/notes/eight/"></a> </li>
</ul>
```

### Using `{% for p in taglist | sort(false, false, 'data.title') %}`
```html
<h1>Étiquette: « all »</h1>
<ul>
  <li data-tags=""><a href="/notes/">All notes</a> </li>
  <li data-tags="notes,animal"><a href="/notes/five/">FiVe</a> <span class="small">| <time>Wed May 14 2003 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,vegetable"><a href="/notes/four/">FoUr</a> <span class="small">| <time>Sun Apr 13 2014 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,mineral"><a href="/notes/one/">oNe</a> <span class="small">| <time>Fri Jan 30 2009 16:00:00 GMT-0800 (Pacific Standard Time)</time></span></li>
  <li data-tags=""><a href="/tags/all/">Tag</a> </li>
  <li data-tags="notes,animal,mineral"><a href="/notes/three/">ThReE</a> </li>
  <li data-tags="notes,vegetable,til"><a href="/notes/two/">tWo</a> <span class="small">| <time>Wed Sep 18 2019 17:00:00 GMT-0700 (Pacific Daylight Saving Time)</time></span></li>
  <li data-tags="notes,til"><a href="/notes/six/"></a> </li>
  <li data-tags="notes"><a href="/notes/seven/"></a> </li>
  <li data-tags="notes"><a href="/notes/eight/"></a> </li>
</ul>
```

**Reply by Julianoe** [reply]
For the first point I found the culprit!

My collection was created as follows:

```js
    eleventyConfig.addCollection("notes", function (collection) {
        return collection.getFilteredByGlob(["notes/**/*.md", "index.md"]);
    });
```

Removing the reference to `index.md` in `getFilteredByGlob` OR adding a `title:` to that index file allowed the notes to sort themselves properly.

**Reply by Julianoe** [reply]
For the second point I spent 1 hour conducting tests and I don't get it. I still receive the error:

```bash
[11ty] Original error stack trace: Template render error: (./tags.njk)
[11ty]   TypeError: Cannot read properties of undefined (reading 'title')`
```

even though I tested it with the exact same code for tags.njk and sortByTitle filter, and **only 1 note** in my notes folder that looks like this:
If I delete the tags, it renders without error (normal) but with tags in I receive error above about reading title. I should add that in .eleventy.js the title is only referred to by the function creating the sortByTitle.

```markdown
---
title: "Test"
id: 20210414054101
date: 2020-04-17 20:20:02
tags:
    - linux
    - hacking
---


Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Duis autem vel eum iriure dolor in hendrerit in vulputate velit esse molestie consequat, vel illum dolore eu feugiat nulla facilisis at vero eros et accumsan et iusto odio dignissim qui blandit praesent luptatum zzril delenit augue duis dolore te feugait nulla facilisi. 
```

Is there something I'm missing about how to format a markdown file 🤔
I don't understand from where this error could come

**Reply by pdehaan** [reply]
I'm at a loss.
Unless there is a weird tab character in the YAML front matter or something, I'm not sure what the issue could be.
I copy/pasted your snippets above into new files locally and things seem to work in the simple test case.

Maybe try that. Start a new project with minimal files and see if you can reproduce the error just with 1-2 notes and that tags.njk file and see if you get the same error. If so, push the simplified test case to GitHub and I can clone it and take a look. If you can't reproduce the error with a simple test case, it's something else in your code/repo causing the error.
Otherwise I'm not sure what else to try without seeing the code/repo and trying locally.

**Reply by pdehaan** [reply]
Here's my sample repo: https://github.com/pdehaan/11ty-2557

**Reply by Julianoe** [reply]
Got it! wow.

One of my YAML frontmatter (among 500 notes) looked like this:
```
---
title: "📸 Scanner Epson Perfection V39"
description: 
date: 2022-07-21 20:18:38
cover:
categories:
tags:
    -
---
```
The empty tag entry caused one of the notes to return `null ` as tags value and that messed up with the code.
The debug mode did not help to debug this; I had to manually check each individual file.

Golden rule folks: your YAML should not contain on once of error!

**Reply by pdehaan** [reply]
For what it's worth, that's 💯 valid YAML; it just happens to return an array with `null` as the only value:

```liquid
---
# src/notes/bad.md (liquid)
title: "📸 Scanner Epson Perfection V39"
description: 
date: 2022-07-21 20:18:38
cover:
categories:
tags:
    -
---

title={{ title }}
tags={{ tags | json }}
```
### OUTPUT
```html
<p>title=📸 Scanner Epson Perfection V39
tags=[null]</p>
```

But I find that to be one of the harder challenges w/ bigger sites. Validating data, and finding the one "malformed" needle in a haystack of posts.

**Reply by pdehaan** [reply]
A bit messy, but here's an example of building a rough schema validator for front matter values using **ajv** and a custom collection: https://github.com/pdehaan/11ty-2557 (look in the .eleventy.js and ./schemas/ directory).

```sh
> eleventy

[./src/notes/bad.md] frontmatter/tags/0 must be string
[./src/notes/note3.md] Unknown schema: 'tangerine'
[./src/notes/note2.md] frontmatter must have required property 'title'

[11ty] Wrote 9 files in 0.08 seconds (v1.0.2)
```

**Reply by Julianoe** [reply]
Thanks, that's really interesting and could show very useful once the site is rolling with potentially hundreds new publications per year
